va/caa.go

```diff
@@ -40,16 +49,164 @@ func (va *ValidationAuthorityImpl) IsCAAValid(ctx context.Context, req *vapb.IsC
+	if features.Get().RVACAARechecking {
+		if remoteVACount := len(va.remoteVAs); remoteVACount > 0 {
+			remoteCAAResults = make(chan *remoteVAResult, remoteVACount)
+			go va.performRemoteCAARecheck(ctx, req, remoteCAAResults)
+		}
+	}
+
+	checkResult := "success"
+	pbProblem := &corepb.ProblemDetails{}
+	prob := va.checkCAA(ctx, acmeID, params)
+	localRecheckLatency := time.Since(checkStartTime)
+	if prob != nil {
 		detail := fmt.Sprintf("While processing CAA for %s: %s", req.Domain, prob.Detail)
-		return &vapb.IsCAAValidResponse{
-			Problem: &corepb.ProblemDetails{
-				ProblemType: string(prob.Type),
-				Detail:      replaceInvalidUTF8([]byte(detail)),
-			},
-		}, nil
-	}
-	return &vapb.IsCAAValidResponse{}, nil
+		pbProblem.ProblemType = string(prob.Type)
+		pbProblem.Detail = replaceInvalidUTF8([]byte(detail))
+		checkResult = "failure"
+	} else if remoteCAAResults != nil {
+		if !features.Get().EnforceMultiVA && features.Get().MultiVAFullResults {
+			// If we're not going to enforce multi VA but we are logging the
+			// differentials then collect and log the remote results in a separate go
+			// routine to avoid blocking the primary VA.
+			go func() {
+				_ = va.processRemoteCAARecheckResultsOuter(
+					req.Domain,
+					req.AccountURIID,
+					string(validationMethod),
+					prob,
+					remoteCAAResults)
+			}()
+			// Since prob was nil and we're not enforcing the results from
+			// `processRemoteCAARecheckResultsOuter` set the challenge status to
+			// valid so the caaRecheckTime metrics increment has the correct
+			// result label.
+		} else if features.Get().EnforceMultiVA {
+			remoteProb := va.processRemoteCAARecheckResultsOuter(
+				req.Domain,
+				req.AccountURIID,
+				string(validationMethod),
+				prob,
+				remoteCAAResults)
+
+			// If the remote result was a non-nil problem then fail the CAA check
+			if remoteProb != nil {
+				prob = remoteProb
+				pbProblem.ProblemType = string(prob.Type)
+				pbProblem.Detail = replaceInvalidUTF8([]byte(prob.Detail))
+				checkResult = "failure"
+				va.log.Infof("CAA check failed due to remote failures: identifier=%v err=%s",
+					req.Domain, remoteProb)
+				va.metrics.remoteCAARecheckFailures.Inc()
+			}
+		}
+	}
+
+	if remoteVACount := len(va.remoteVAs); remoteVACount > 0 {
+		recheckLatency := time.Since(checkStartTime)
+		va.metrics.localValidationTime.With(prometheus.Labels{
+			"type":   req.ValidationMethod,
+			"result": checkResult,
+		}).Observe(localRecheckLatency.Seconds())
+		if prob == nil {
+			va.metrics.caaRecheckTime.With(prometheus.Labels{
+				"type":         req.ValidationMethod,
+				"result":       checkResult,
+				"problem_type": "",
+			}).Observe(recheckLatency.Seconds())
+		} else {
+			va.metrics.caaRecheckTime.With(prometheus.Labels{
+				"type":         req.ValidationMethod,
+				"result":       checkResult,
+				"problem_type": string(prob.Type),
+			}).Observe(recheckLatency.Seconds())
+		}
+	}
+
```

Comment:
This is a great idea for a followup PR.

Reply:
The authz ID is not currently passed to [IsCAAValidRequest](https://github.com/letsencrypt/boulder/blob/a9a87cd4a897d5fcaf79a1f22351398448565ac8/va/proto/va.proto#L16-L21) unlike `PerformValidationRequest` ([[1]](https://github.com/letsencrypt/boulder/blob/a9a87cd4a897d5fcaf79a1f22351398448565ac8/va/proto/va.proto#L28-L32) and [[2]](https://github.com/letsencrypt/boulder/blob/a9a87cd4a897d5fcaf79a1f22351398448565ac8/va/va.go#L672-L676)) which allows logging enough data to determine which sufficiently old authz triggered the CAA check. I think we do want that, but it would need to be done in a separate PR, get deployed to staging/prod, then accessed here once the protobuf change has been successfully rolled out.

Reply:
From sprint planning, I'll make the protobuf change as a followup PR.